src/collector/top_score_collector.rs

```diff
@@ -736,6 +722,64 @@ impl SegmentCollector for TopScoreSegmentCollector {
     }
 }
 
+pub struct TopNComputer<Score, DocId> {
+    buffer: Vec<ComparableDoc<Score, DocId>>,
+    max_size: usize,
+    pub threshold: Option<Score>,
+}
+
+impl<Score, DocId> TopNComputer<Score, DocId>
+where
+    Score: PartialOrd + Clone,
+    DocId: Ord + Clone,
+{
+    pub fn new(top_n: usize) -> Self {
+        let max_size = top_n.max(1) * 2;
+        TopNComputer {
+            buffer: Vec::with_capacity(max_size),
+            max_size,
+            threshold: None,
+        }
+    }
+
+    #[inline]
+    pub(crate) fn push(&mut self, doc: ComparableDoc<Score, DocId>) {
+        if let Some(last_median) = self.threshold.clone() {
```

Comment:
If `new` were to take a `initial_thresold` argument to initialize `self.threshold` to e.g. `Score::MIN`, could the `Option` layer be avoided completely?

Reply:
Yes, but it's already a generic without that function on the caller side

Reply:
I am sorry, but I just don't understand. I meant to have

```rust
pub struct TopNComputer<Score, DocId> {
    buffer: Vec<ComparableDoc<Score, DocId>>,
    top_n: usize,
    pub(crate) threshold: Score,
}
```

so that the `if let Some(threshold) = self.threshold.clone()` can be omitted, i.e. branching on the `Option` tag is removed.

Reply:
`Score` is a generic, we don't have access to `Score::MIN`

Reply:
Not in `TopNComputer`, yes. But the caller of `TopNComputer::new` has access and hence could pass this value on like

```rust
let mut top_collector = TopNComputer::new(self.limit + self.offset, Score::MIN);
```

Reply:
No, as I wrote before it's not available on the caller side, there's just `T`

Reply:
Okay, then I this was what I did not parse from

> Yes, but it's already a generic without that function on the caller side